addons/mail/tools/guard_discuss_access.py

```diff
@@ -0,0 +1,86 @@
+from urllib.parse import urlsplit
+from functools import wraps
+from inspect import Parameter, signature
+from werkzeug.exceptions import NotFound
+
+from odoo.tools import consteq
+from odoo.http import request
+from odoo.addons.bus.websocket import wsrequest
+
+
+is_allowed_sentinel = object()
+
+
+def add_guest_to_context(func):
+    """Decorate a function to extract the guest from the request.
+    The guest is then available on the context of the current request. In case
+    of cross origin request, the `guest_token` passed in the request params is
+    used to authenticate the guest. The authentication cookie is used otherwise.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        req = request or wsrequest
+        has_user = False
+        origin_url = urlsplit(req.httprequest.headers.get("origin", req.httprequest.headers.get("referer")))
+        if (
+            origin_url.netloc == req.httprequest.headers.get("host") and origin_url.scheme == req.httprequest.scheme
+        ) or req.httprequest.environ.get("HTTP_SEC_FETCH_DEST") == "document":
+            has_user = not req.env.user._is_public()
```

Comment:
inverse condition: if there is a user, the origin must match
we should not allow a different origin if there is a guest (but also logged as admin)